AK/CPUFeature.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2024, Dan Klishch <danilklishch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/EnumBits.h>
+#include <AK/Types.h>
+
+namespace AK {
+
+enum class CPUFeature : u64 {
+    None = 0ULL,
+    Invalid = 1ULL << 63,
+
+#if !defined(KERNEL) && ARCH(X86_64)
+#    define AK_CAN_CODEGEN_FOR_X86_SSE42 1
+    X86_SSE42 = 1 << 0,
+#    define AK_CAN_CODEGEN_FOR_X86_AVX 1
+    X86_AVX = 1 << 1,
+#    define AK_CAN_CODEGEN_FOR_X86_SHA 1
+    X86_SHA = 1 << 2,
+#else
+#    define AK_CAN_CODEGEN_FOR_X86_SSE42 0
+    X86_SSE42 = Invalid,
+#    define AK_CAN_CODEGEN_FOR_X86_AVX 0
+    X86_AVX = Invalid,
+#    define AK_CAN_CODEGEN_FOR_X86_SHA 0
+    X86_SHA = Invalid,
+#endif
+};
+
+#if ARCH(X86_64)
+struct CPUIDResult {
+    u32 eax { 0 };
+    u32 ebx { 0 };
+    u32 ecx { 0 };
+    u32 edx { 0 };
+};
+
+inline CPUIDResult cpuid(u32 leaf, u32 subleaf)
+{
+    CPUIDResult result;
+    asm("cpuid"
+        : "=a"(result.eax), "=b"(result.ebx), "=c"(result.ecx), "=d"(result.edx)
+        : "0"(leaf), "2"(subleaf));
+    return result;
+}
+#endif
+
+AK_ENUM_BITWISE_OPERATORS(CPUFeature);
+
+inline CPUFeature detect_cpu_features()
+{
+    static CPUFeature const s_cached_features = [] {
+        CPUFeature result = CPUFeature::None;
+
+#if ARCH(X86_64)
+        u32 max_leaf = cpuid(0, 0).eax;
+        [[maybe_unused]] auto cpuid1 = max_leaf >= 1 ? cpuid(1, 0) : CPUIDResult {};
+        [[maybe_unused]] auto cpuid7 = max_leaf >= 7 ? cpuid(7, 0) : CPUIDResult {};
+
+#    if AK_CAN_CODEGEN_FOR_X86_SSE42
+        if (cpuid1.ecx >> 20 & 1)
+            result |= CPUFeature::X86_SSE42;
+#    endif
+#    if AK_CAN_CODEGEN_FOR_X86_AVX
+        if (cpuid1.ecx >> 28 & 1)
+            result |= CPUFeature::X86_AVX;
+#    endif
+#    if AK_CAN_CODEGEN_FOR_X86_SHA
+        if (cpuid7.ebx >> 29 & 1)
+            result |= CPUFeature::X86_SHA;
+#    endif
+#endif
+
+        return result;
+    }();
+    return s_cached_features;
+}
+
+#define AK_UPDATE_SELECTION_IF_FEATURE_SUPPORTED(feature, template_function) \
+    if constexpr (((feature) & CPUFeature::Invalid) == CPUFeature::None) {   \
+        if (has_flag(detect_cpu_features(), (feature)))                      \
+            result = template_function<(feature)>;                           \
```

Comment:
This macro seems pretty janky to me.

1. It just silently assumes there's a symbol `result` available which it overwrites
2. It requires per-feature function implementations to be in in a template specialization, which seems needlessly clever
3. It for some reason supports passing in `CPUFeature::Invalid` in the (compile-time static) first `feature` parameter

Reply:
While I don't like the fact I have to add this macro too, I can definitely defend other design decisions (expect for implied `result` variable name, that's just dumb of me).

The commit in question actually does 3 different things (which I called a "framework" and haven't bothered to split up). The first one is removing the current and any possible future ifdef soup in files that try to choose from >1 different implementations. I hope you agree that `#if AK_CAN_CODEGEN_FOR_X86_SHA && AK_CAN_CODEGEN_FOR_X86_SSE42` is much better than some random `#if (ARCH(I386) || ARCH(X86_64))`.

The second one is disambiguating between implementations using template arguments instead of name suffixes. The motivation here comes from my ever-work-in-progress FFT multiplication (https://github.com/DanShaders/serenity/tree/fft-multiplication). There I have multiple layers of functions in the vectorized "kernel" with most of the code in the higher levels being generic but still sensitive to virtual calls that will come if I use target clones. If I use suffixes, I will have to either pass function pointers or just copy-paste code, while with templates, having devirtualized "target clones" is as easy as
```cpp
template TARGET_AVX2 void nttf<SIMD::AVX2, 2>(size_t, size_t, u64*, size_t, size_t, u64*, u64*, u64*, NativeWord*);
template TARGET_AVX2 void nttr<SIMD::AVX2>(size_t, size_t, u64*, u64*, u64*, u64*);
template TARGET_AVX2 void nttm<SIMD::AVX2>(size_t, u64*);
```

And finally, the third one is writing ifunc resolvers (or, currently, because of Apple Clang, static function pointer constant) dispatchers using the macro in question. Well, without it, the dispatcher looks like
```cpp
static decltype(transform_impl<CPUFeature::None>)* const transform_dispatched = [] {
    [[maybe_unused]] auto result = transform_impl<CPUFeature::None>;
    [[maybe_unused]] auto features = detect_cpu_features();

#if AK_CAN_CODEGEN_FOR_X86_SHA && AK_CAN_CODEGEN_FOR_X86_SSE42
    if (has_flag(features, CPUFeature::X86_SSE42 | CPUFeature::X86_SHA))
        result = transform_impl<CPUFeature::X86_SSE42 | CPUFeature::X86_SHA>;
#endif

    return result;
}();
``` 

which seems overly verbose. And finally-finally, answering one of the actual questions, `if constexpr` in the macro is effectively a `#if AK_CAN_CODEGEN_FOR_X86_SHA && AK_CAN_CODEGEN_FOR_X86_SSE42` check from the macro-less dispatcher.